rfd/0159-webasset-cache.md

```diff
@@ -0,0 +1,164 @@
+
+Depending how we end up syncing the webassets to each available auth, we may
+also add another message of `SendWebasset` along the same lines as Get above.
+
+#### Syncing webassets to auth
+
+Because there can be multiple auth servers in a cluster, each proxy will need to
+report it's assets to every auth server in the cluster. We can do this by
+INSERT_HOW_TO_DO_THIS. 
+The proxy should listen to auth heartbeats and be able to ask "have I sent this file
+to this auth recently" and if not, send it over. These will be stored by fileName->fileBytes. If a file
+exists in the cache already when it's sent (two of the same version), we can
+check the bytes to see if they match. If they don't match, we should remove the
+file from the cache as we don't know which proxy is sending the correct file.
+Worst case scenario, the file isn't found and the proxy returns a `404` similar
+to what it does now. Best case scenario, we prevented some mismatch 
+file that would break code, or perhaps a malicious file from being stored in the cache. 
+We can store removed fileNames in a separate list to ensure a removed file isn't just 
+added again later after removal.
+
+#### Alternatives to syncing with auth
+An alternative that was suggested to store the web assets reported to auth in
+the backend. That way, if auth didn't have the asset the proxy was asking for
+(in a multi auth situation) then the backend could be the final fallback but
+some of the webassets are too large to store in the backend per item (400kb for
+dynamoDB for example). So we scrapped this idea.
+
+
+#### Serving the missing webassets
+Currently, our static file serving looks [like this](https://github.com/gravitational/teleport/blob/master/lib/web/apiserver.go#L513C4-L518)
+```go
+if strings.HasPrefix(r.URL.Path, "/web/app") {
+    fs := http.FileServer(cfg.StaticFS)
+
+    fs = makeGzipHandler(fs)
+    fs = makeCacheHandler(fs)
+
+    http.StripPrefix("/web", fs).ServeHTTP(w, r)
+}
+```
+
+nothing too complicated besides setting some headers, compression, and then
+serving whatever file is requested. The change would be instead of serving,
+first check if the file exists in the embedded file system. If it does, send as usual.
+If it doesn't, we can fetch the requested file over gRPC and serve the file bytes 
+instead with something like this
+```go
+if strings.HasPrefix(r.URL.Path, "/web/app") {
+	// do everything the same if the file exists
+	if (fileExists) {
+		fs := http.FileServer(cfg.StaticFS)
+
+		fs = makeGzipHandler(fs)
+		fs = makeCacheHandler(fs)
+
+		http.StripPrefix("/web", fs).ServeHTTP(w, r)
+        return
+	}
+    res, err := h.auth.proxyClient.GetWebasset(cfg.Context, &webassetcache.GetWebassetRequest{
+		Name: r.URL.Path[len("/web/app/"):],
+	})
+	// detect content type
+	// compress
+	// store the file in this proxy's webasset cache to prevent future requests
+	http.ServeContent(w, r, "example.txt", time.Now(), compressedFileHere)
+}
+```
+
+
+### Security Concerns
```

Comment:
I'm not too aware of what a proxy compromise looks like - but - I'd assume some of the concerns about a proxy compromise would be a case where a proxy that is serving traffic is compromised. I wonder what the current blast radius of a proxy token being compromised would be (e.g what special actions a proxy is allowed to complete without any knowledge of a users authentication) and how those compare to being able to inject malicious content.

Perhaps a little complex, but maybe we should consider (and potentially reject) the idea of us signing individual webassets or a webassets bundle during the release process and therefore providing uncompromised auth servers/proxies a way to reject maliciously injected content.

Reply:
The proxy already serves web content. If a proxy is compromised they can serve malicious content without the need of storing it in a webasset cache. I will see if there is something that can be done about this and add to the RFD.

Reply:
Ah I see what you mean. Good point! Let me see what I can figure out.

Reply:
Can't we just embed version-1 assets into the new proxy version -  this solves the fact that the newest version can handle the older version assets.
For the older version, it can route/cache the request to the new proxy by watching proxies from auth server and retrying until it connects to the newest version - or add some mechanism do dial via auth
Allowing proxies to push stuff into auth will be a considerable security problem since it's allowed even outside of upgrade windows.

Reply:
The web assets can change even with patch releases so we can't arbitrarily decide which version/minor/patch to store. Someone could be upgrading from 13.1 to 14.2.2 for example.

Reply:
Oh that's correct. 
Then the only viable option is for each proxy to connect to all other proxies versions and pull their webassets. At least we guarantee that the proxies are reachable from the public domain and they aren't just joining tokens or fake proxies.